Llm d http route reconciler

andresllh opened the pull request:
**What this PR does / why we need it**:
Depends on https://github.com/kserve/kserve/pull/4614 and https://github.com/kserve/kserve/pull/4596
This PR adds the capability to reconcile routes for llm ISVC.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.
`make test`

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

VedantMahabaleshwarkar committed:
update CRDs for LWS based multi node support

Co-authored-by: Bartosz Majsak <bartosz.majsak@gmail.com>
Co-authored-by: Pierangelo Di Pilato <pierdipi@redhat.com>
Signed-off-by: Vedant Mahabaleshwarkar <vmahabal@redhat.com>

andresllh committed:
Adding llmd scheduler and its dependencies.

Co-authored-by: Bartosz Majsak <bartosz.majsak@gmail.com>
Co-authored-by: Pierangelo Di Pilato <pierdipi@redhat.com>
Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Merge branch 'master' into llm-d-scheduler

andresllh committed:
Merge branch 'master' into llm-d-scheduler

andresllh committed:
Ran make precommit

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Adding reconciler for router plus dependencies.

Co-authored-by: Bartosz Majsak <bartosz.majsak@gmail.com>
Co-authored-by: Pierangelo Di Pilato <pierdipi@redhat.com>
Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Merge branch 'master' of github.com:kserve/kserve into llm-d-http-route-reconciler

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Retrigger CI

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Ran make precommit

Signed-off-by: Andres Llausas <allausas@redhat.com>

hdefazio commented on pkg/apis/serving/v1alpha1/llm_inference_service_lifecycle.go:
```diff
@@ -1,5 +1,9 @@
 /*
+<<<<<<< HEAD
```
merge conflict?

andresllh replied:
Good catch

andresllh committed:
Fix merge conflict

Signed-off-by: Andres Llausas <allausas@redhat.com>

gavrissh commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,214 @@
+	referencedRoutes := make([]*gwapiv1.HTTPRoute, 0, len(llmSvc.Spec.Router.Route.HTTP.Refs))
+	for _, routeRef := range llmSvc.Spec.Router.Route.HTTP.Refs {
+		route := &gwapiv1.HTTPRoute{}
+		if err := r.Client.Get(ctx, types.NamespacedName{Namespace: routeRef.Name, Name: llmSvc.GetNamespace()}, route); err != nil {
```
Incorrect assignment?

```suggestion
		if err := r.Client.Get(ctx, types.NamespacedName{Namespace: llmSvc.GetNamespace(), Name: routeRef.Name}, route); err != nil {
```

andresllh replied:
Done

gavrissh commented on pkg/controller/v1alpha1/llmisvc/router_discovery.go:
```diff
@@ -0,0 +1,203 @@
+func DiscoverGateways(ctx context.Context, c client.Client, route *gwapiv1.HTTPRoute) ([]resolvedGateway, error) {
+	gateways := make([]resolvedGateway, 0)
+	for _, parentRef := range route.Spec.ParentRefs {
+		ns := ptr.Deref((&parentRef).Namespace, gwapiv1.Namespace(route.Namespace))
```
Is parentRef a pointer? If not, this (&parentRef) is unnecessary and confusing?

andresllh replied:
Done

gavrissh commented on pkg/controller/v1alpha1/llmisvc/router_discovery.go:
```diff
@@ -0,0 +1,203 @@
+		}
+	}
+
+	return &gateway.Spec.Listeners[0]
```
Is this expected to fall back to the default if sectionName not found?

andresllh replied:
Spec and Listeners can't be empty, so it'll default to the first one in the list.

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router_discovery.go:
```diff
@@ -0,0 +1,203 @@
+	parentRef    gwapiv1.ParentReference
+}
+
+func DiscoverGateways(ctx context.Context, c client.Client, route *gwapiv1.HTTPRoute) ([]resolvedGateway, error) {
```
this is an internal function we do not need to export

andresllh replied:
Done

yuzisun commented on pkg/controller/v1alpha1/llmisvc/scheduler.go:
```diff
@@ -0,0 +1,481 @@
+func schedulerConfigText(llmSvc *v1alpha1.LLMInferenceService) string {
+	switch {
+	case llmSvc.Spec.Prefill != nil:
+		return `
```
why do we return the string here?

andresllh replied:
the string gets appended to container args, which is a string

yuzisun commented on pkg/controller/v1alpha1/llmisvc/scheduler.go:
```diff
@@ -0,0 +1,481 @@
+
+			d.Spec.Template.Spec.Containers[i].Args = append(d.Spec.Template.Spec.Containers[i].Args,
+				"--configText",
+				schedulerConfigText(llmSvc),
```
Can we refer to the `EndpointPickerConfig` resource name instead of passing the text ?

pierDipi replied:
it's not a real CRD

johnugeorge replied:
Can't it be read into a type
https://github.com/kubeflow/katib/blob/master/pkg/util/v1beta1/katibconfig/config.go#L149-L160

yuzisun commented on pkg/controller/v1alpha1/llmisvc/scheduler.go:
```diff
@@ -0,0 +1,481 @@
+			}
+
+			if slices.Contains(d.Spec.Template.Spec.Containers[i].Args, "--configText") ||
+				slices.Contains(d.Spec.Template.Spec.Containers[i].Args, "-configText") ||
```
why we use two different arg naming conventions here?

pierDipi replied:
the config can be passed and overridden by users via configText or configFile flag (and in Go it can be passed with both single and double dash), so if it's overridden we don't use our default configuration

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,214 @@
+		return fmt.Errorf("failed to reconcile HTTP routes: %w", err)
+	}
+
+	if err := r.reconcileIstioDestinationRules(ctx, llmSvc); err != nil {
```
This makes Istio as a hard dependency for llmisvc

andresllh replied:
I've removed the istio hard dependency from this PR. It will be reworked and added as a separate PR.

yuzisun commented on pkg/controller/v1alpha1/llmisvc/workload_tls_self_signed_istio.go:
```diff
@@ -0,0 +1,227 @@
+	return dr, nil
+}
+
+func (r *LLMISVCReconciler) getIstioShadowInferencePoolService(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService) (*corev1.Service, error) {
```
Can you add comment to explain what's shadow InferencePool serivce?

andresllh replied:
@pierDipi @bartoszmajsak @Jooho

pierDipi replied:
Given the other comment on moving away from Istio specific APIs, do we still need to describe it?

Generally, this is an Istio implementation detail for the InferencePool API, it creates this service to add it to the "service registry"

andresllh replied:
I've removed the Istio implementation from this PR, given that we want to move away from the hard dependency. I will rework it and add it as another PR later on.

yuzisun commented:
I have a bit concerns here that the implementation is tightly coupled with Istio API, ideally we should only use Gateway API so it is agnostic of the gateway implementation.

yuzisun commented on pkg/controller/v1alpha1/llmisvc/scheduler.go:
```diff
@@ -0,0 +1,481 @@
+	"github.com/kserve/kserve/pkg/apis/serving/v1alpha1"
+)
+
+func (r *LLMISVCReconciler) reconcileScheduler(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService) error {
```
Please add tests for these new files

andresllh replied:
the controller tests are rather large and there were a bunch of additions made to the testing package. Would it make more sense to add all tests as part of another PR?

andresllh replied:
https://github.com/kserve/kserve/issues/4629

terrytangyuan commented:
> I have a bit concerns here that the implementation is tightly coupled with Istio API, ideally we should only use Gateway API so it is agnostic of the gateway implementation.

Agreed. @andresllh Let's revisit the implementation in this PR to make sure to not depend on Istio.

terrytangyuan commented on pkg/controller/v1alpha1/llmisvc/controller.go:
```diff
@@ -50,3 +59,49 @@ func (r *LLMISVCReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		IstioGatewayControllerNames: []string{
+			"istio.io/gateway-controller",
+			"istio.io/unmanaged-gateway",
+			"openshift.io/gateway-controller",
```
These are hard-coded for Istio and OpenShift

andresllh replied:
I've removed any Istio dependencies from this PR. The logic will be modified and added as another PR later on.

andresllh committed:
Merge branch 'master' into llm-d-http-route-reconciler

andresllh committed:
Addressing PR comments. Removed Istio code.

Co-authored-by: Bartosz Majsak <bartosz.majsak@gmail.com>
Co-authored-by: Pierangelo Di Pilato <pierdipi@redhat.com>
Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Ran precommit

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh commented:
Will be rebasing tomorrow morning @johnugeorge . For now it's ready for review once more. I've removed the Istio hard dependency from this PR so that it can be reworked and added in another PR later.

andresllh committed:
Retrigger CI

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Address scheduler.go comments

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Ran make precommit

Signed-off-by: Andres Llausas <allausas@redhat.com>

yuzisun committed:
Merge branch 'master' into llm-d-http-route-reconciler

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Update workload_single_node.go

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

andresllh committed:
Removed remaining istio references.

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Merge branch 'master' into llm-d-http-route-reconciler

andresllh committed:
Ran precommit

Signed-off-by: Andres Llausas <allausas@redhat.com>

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,209 @@
+	}
+
+	if llmSvc.Spec.Router != nil && llmSvc.Spec.Router.Route != nil && llmSvc.Spec.Router.Route.HTTP.Spec != nil {
+		httpRoute.Spec = *llmSvc.Spec.Router.Route.HTTP.Spec.DeepCopy()
```
I think we should create the HTTPRoute for user as kserve knows what routing rule to create instead of providing the spec

andresllh replied:
A user can enable the route by just using `route: {}`  which will have the spec be injected using the presets, but it is still overridable if a user wants to override the preset.

Route preset:
https://github.com/kserve/kserve/blob/7794ff31bd39cab87e4d0664ee68df1af75ab395/pkg/controller/v1alpha1/llmisvc/config_merge.go#L101-L103

yuzisun commented on pkg/controller/v1alpha1/llmisvc/fixture/gwapi_builders.go:
```diff
@@ -0,0 +1,364 @@
+			Name: name,
+		},
+		Spec: gwapiv1.GatewaySpec{
+			GatewayClassName: defaultGatewayClass,
```
there is a gatewayclass in the configmap, we can read from there. 
https://github.com/kserve/kserve/blob/master/config/configmap/inferenceservice.yaml#L293

andresllh replied:
Even for test fixture?

yuzisun replied:
nvm, did not realize these are fixtures based on the name of the file.

andresllh committed:
Merge branch 'master' into llm-d-http-route-reconciler

Jooho commented:
@yuzisun Could you please take another look when you have a chance?

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router_discovery.go:
```diff
@@ -0,0 +1,203 @@
+	return urls, nil
+}
+
+func extractRoutePath(route *gwapiv1.HTTPRoute) string {
```
Can you comment on what is function doing?

andresllh replied:
Done

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router_discovery.go:
```diff
@@ -0,0 +1,203 @@
+	parentRef    gwapiv1.ParentReference
+}
+
+func discoverGateways(ctx context.Context, c client.Client, route *gwapiv1.HTTPRoute) ([]resolvedGateway, error) {
```
It would be helpful to add comments and provide the examples what these functions are doing

andresllh replied:
Done

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,209 @@
+
+	var urls []*apis.URL
+	for _, route := range routes {
+		discoverURL, err := DiscoverURLs(ctx, r.Client, route)
```
discoverURL is an array, so should be using plural

andresllh replied:
Done

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,209 @@
+
+	expectedHTTPRoute := r.expectedHTTPRoute(ctx, llmSvc)
+
+	// TODO should we remove "llmSvc.Spec.Router.Route.HTTP == nil" from the condition below so that a non nil Route meeans "all type of routes are enabled"?
```
```suggestion
	// TODO should we remove "llmSvc.Spec.Router.Route.HTTP == nil" from the condition below so that a non nil Route means "all type of routes are enabled"?
```

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,209 @@
+
+	externalURLs := FilterExternalURLs(urls)
+	if len(externalURLs) == 0 {
+		logger.Info("no public URL discovered")
```
should this be an error instead ? the llmisvc should always have a url

andresllh replied:
this will only expose the first available public url, the private url should always be there. I guess the question is, if there are no public urls, do we want to have the privateURL listed in the llmSvc.Status.URL or just keep it blank? All available URLs will still be listed under llmSvc.Status.Addresses both public and private.
Any thoughts? @yuzisun @pierDipi @bartoszmajsak @Jooho

Jooho replied:
I think externalURL should be treated as info rather than an error, because there are several situations where it may be unset either unintentionally or by design.
For example, if there is no external load balancer setup, or if it’s not configured properly, externalURL may be missing — and this could be intentional. Another case is when a user wants the service to remain internal and does not want it exposed.
Therefore, in these scenarios, logging it as informational rather than an error makes more sense.

yuzisun replied:
Is there an error condition in the status to indicate that the load balancer is not configured properly?

Jooho replied:
By default, the following logic treats the absence of any external addresses in the Gateway’s status as an error.
If the configuration is correct, this field should be properly populated with the expected address.

https://github.com/kserve/kserve/pull/4617/files/1e5446ab1a187c4d236a1a03e2006851c96ef4b6#diff-39e77924bad68b0a93b51989b48b2dac5e69810aa61ee13382540d77799163eaR77

yuzisun commented on pkg/controller/v1alpha1/llmisvc/router_discovery.go:
```diff
@@ -0,0 +1,203 @@
+func combineIntoURLs(hostnames []string, scheme string, port gwapiv1.PortNumber, path string) ([]*apis.URL, error) {
+	urls := make([]*apis.URL, 0, len(hostnames))
+
+	sortedHostnames := make([]string, len(hostnames))
```
Should we explicitly configure the primary host instead of relying on the sorting ?

bartoszmajsak replied:
Could you elaborate on the use case here? I am curious where that would help. 

What would be the primary hostname in this case - an ingress domain?

yuzisun replied:
Since we are only showing one URL in the status, should that be an explicitly configured domain instead of relying on sorting logic which is a bit arbitrary.

bartoszmajsak replied:
We can address it as a follow-up. The idea is that this function sorts all addresses to have them stable. Then we find first external address and expose is as `.status.url`, everything goes into `.status.addresses`. Is it fine for you @yuzisun ?

https://github.com/kserve/kserve/blob/b5e3d533eb7e2f5b6a02b9d4c01a6baa00fc0eb0/pkg/controller/v1alpha1/llmisvc/router.go#L165-L182

bartoszmajsak replied:
We don't need to sort here, as it's done in the caller, I will clean it up.

andresllh committed:
Apply suggestion from @yuzisun

Co-authored-by: Dan Sun <dsun20@bloomberg.net>
Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Addressed Dan comments

Signed-off-by: Andres Llausas <allausas@redhat.com>

andresllh committed:
Merge branch 'master' into llm-d-http-route-reconciler

sivanantha321 commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,210 @@
+
+	"k8s.io/utils/ptr"
+
+	"k8s.io/apimachinery/pkg/api/equality"
```
nit: reformat the imports

andresllh replied:
Done

sivanantha321 commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,210 @@
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/utils/ptr"
```
nit: reformat imports

andresllh replied:
Done

sivanantha321 commented on pkg/controller/v1alpha1/llmisvc/router.go:
```diff
@@ -0,0 +1,210 @@
+
+	var urls []*apis.URL
+	for _, route := range routes {
+		discoverURLs, err := DiscoverURLs(ctx, r.Client, route)
```
Is there a possibility for duplicate urls ?

yuzisun committed:
Merge branch 'master' into llm-d-http-route-reconciler

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Update controller.go

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Update controller_types.go

Signed-off-by: Dan Sun <dsun20@bloomberg.net>
